File: love/game/turn_manager.lua
-- game/turn_manager.lua
-- Orchestrates the turn cycle: deal → resolve 3 → carry over → repeat.
-- This module manages the flow of a Scoundrel game but delegates rendering
-- and input to the game_play state.

local Class      = require("lib.class")
local Deck       = require("entities.deck")
local WeaponSlot = require("entities.weapon_slot")
local Room       = require("game.room")
local Combat     = require("game.combat")
local Rules      = require("game.rules")
local Scoring    = require("game.scoring")
local Event      = require("lib.event")
local C          = require("core.constants")

local TurnManager = Class:extend()

function TurnManager:init()
    self.deck = Deck:new()
    self.room = Room:new()
    self.weapon_slot = WeaponSlot:new()

    self.hp = C.START_HP
    self.avoided_last_room = false
    self.last_card = nil -- last card resolved (for scoring)

    -- The carry-over card from the previous room (4th unresolved card)
    self.carry_over = nil
    self.carry_over_slot = nil -- which slot (1-4) it was in

    -- Game flow state
    self.game_over = false
    self.victory = false
    self.score = 0

    -- Track the phase of the current turn
    -- "dealing" → "resolving" → "room_complete" → back to "dealing"
    self.phase = "ready"
end

-- Start a new turn: deal cards into the room.
-- Returns slot_cards (table keyed 1-4) and carry_over_slot (or nil).
-- The carry-over stays in its original slot; new cards fill the empty ones.
function TurnManager:deal_room()
    self.phase = "dealing"
    local slot_cards = {}
    local co_slot = nil

    -- Carry-over stays in its original slot
    if self.carry_over then
        co_slot = self.carry_over_slot
        slot_cards[co_slot] = self.carry_over
        self.carry_over = nil
        self.carry_over_slot = nil
    end

    -- Fill empty slots from the dungeon deck
    for i = 1, 4 do
        if not slot_cards[i] and not self.deck:is_empty() then
            slot_cards[i] = self.deck:draw_card()
        end
    end

    self.room:set_cards(slot_cards)
    self.phase = "resolving"

    Event.emit("room_dealt", { slot_cards = slot_cards, carry_over_slot = co_slot })
    return slot_cards, co_slot
end

-- Can the player avoid the current room?
function TurnManager:can_avoid()
    return Rules.can_avoid_room(self.avoided_last_room)
        and self.room:card_count() == 4  -- can only avoid a full 4-card room
        and self.phase == "resolving"
        and self.room.resolved_count == 0 -- haven't started resolving yet
end

-- Avoid the current room: all 4 cards go to bottom of dungeon.
function TurnManager:avoid_room()
    if not self:can_avoid() then return false end

    local cards = self.room:get_all_cards()
    self.room:clear()
    self.deck:place_at_bottom(cards)
    self.avoided_last_room = true

    Event.emit("room_avoided", { cards = cards })

    -- Don't auto-deal here — the visual layer (game_play) handles dealing
    -- so it can animate the cards moving back to the dungeon first.
    return true
end

-- Resolve a card at the given room slot index.
-- For monsters, `use_weapon` determines barehanded vs weapon combat.
-- Returns a result table describing what happened.
function TurnManager:resolve_card(slot_index, use_weapon)
    if self.phase ~= "resolving" then return nil end
    if not self.room:has_card_at(slot_index) then return nil end

    local card = self.room.slots[slot_index]
    local result = { card = card, slot = slot_index }

    if card.is_potion then
        result = self:resolve_potion(slot_index, card)
    elseif card.is_weapon then
        result = self:resolve_weapon(slot_index, card)
    elseif card.is_monster then
        result = self:resolve_monster(slot_index, card, use_weapon)
    end

    -- Check if room is complete after resolving
    if self.room:is_complete() then
        self:finish_room()
    end

    -- Check for game over
    if self.hp <= 0 then
        self:end_game()
    end

    return result
end

-- Resolve a potion card.
function TurnManager:resolve_potion(slot_index, card)
    local healed = 0
    local wasted = false

    if Rules.can_use_potion(self.room) then
        local old_hp = self.hp
        self.hp = Combat.apply_healing(self.hp, card.value, C.MAX_HP)
        healed = self.hp - old_hp
        self.room.potion_used = true
    else
        -- Second potion this turn: wasted, no healing
        wasted = true
    end

    self.room:resolve_card(slot_index)
    self.last_card = card

    local result = {
        action = "potion",
        card = card,
        slot = slot_index,
        healed = healed,
        wasted = wasted,
    }

    Event.emit("potion_used", result)
    return result
end

-- Resolve a weapon card (equip it).
function TurnManager:resolve_weapon(slot_index, card)
    local old_weapon, old_slain = self.weapon_slot:equip(card)

    self.room:resolve_card(slot_index)
    self.last_card = card

    local result = {
        action = "weapon_equipped",
        card = card,
        slot = slot_index,
        old_weapon = old_weapon,
        old_slain = old_slain,
    }

    Event.emit("weapon_equipped", result)
    return result
end

-- Resolve a monster card (fight it).
function TurnManager:resolve_monster(slot_index, card, use_weapon)
    local damage
    local used_weapon = false

    if use_weapon and Rules.can_use_weapon(self.weapon_slot, card) then
        -- Fight with weapon
        damage = Combat.weapon_damage(card.value, self.weapon_slot:weapon_value())
        self.weapon_slot:record_kill(card)
        used_weapon = true
    else
        -- Fight barehanded
        damage = Combat.barehanded_damage(card.value)
    end

    self.hp = Combat.apply_damage(self.hp, damage)
    self.room:resolve_card(slot_index)
    self.last_card = card

    local result = {
        action = "monster_fought",
        card = card,
        slot = slot_index,
        damage = damage,
        used_weapon = used_weapon,
    }

    Event.emit("monster_fought", result)
    return result
end

-- Called when all 3 cards have been resolved in a room.
function TurnManager:finish_room()
    -- The remaining card carries over to the next room, keeping its slot position
    self.carry_over, self.carry_over_slot = self.room:get_remaining_card()
    self.room:clear()
    self.avoided_last_room = false -- can avoid the next room

    self.phase = "room_complete"
    Event.emit("room_complete", {
        carry_over = self.carry_over,
        carry_over_slot = self.carry_over_slot,
    })

    -- If the deck is empty and there's no carry-over, game is over
    if self.deck:is_empty() and not self.carry_over then
        self:end_game()
    end
end

-- End the game and calculate the score.
function TurnManager:end_game()
    self.game_over = true
    self.victory = self.hp > 0
    self.score = Scoring.calculate(self.hp, self.deck, self.last_card)

    Event.emit("game_over", {
        victory = self.victory,
        score = self.score,
        hp = self.hp,
    })
end

-- Get available actions for a card at the given slot index.
-- Returns a table of action strings the player can take.
function TurnManager:get_actions(slot_index)
    if self.phase ~= "resolving" then return {} end
    if not self.room:has_card_at(slot_index) then return {} end

    local card = self.room.slots[slot_index]
    local actions = {}

    if card.is_potion then
        -- Potions can always be taken (even if wasted)
        table.insert(actions, "use_potion")
    elseif card.is_weapon then
        table.insert(actions, "equip_weapon")
    elseif card.is_monster then
        table.insert(actions, "fight_barehanded")
        if Rules.can_use_weapon(self.weapon_slot, card) then
            table.insert(actions, "fight_with_weapon")
        end
    end

    return actions
end

return TurnManager

File: ebitengine/game/turn_manager.go
package game

import (
	"github.com/thesimpledev/radianthorizon/core"
	"github.com/thesimpledev/radianthorizon/entities"
	"github.com/thesimpledev/radianthorizon/lib"
)

// ResolveResult describes what happened when a card was resolved.
type ResolveResult struct {
	Action     string
	Card       *entities.Card
	Slot       int
	Damage     int
	Healed     int
	Wasted     bool
	UsedWeapon bool
	OldWeapon  *entities.Card
	OldSlain   []*entities.Card
}

// TurnManager orchestrates the turn cycle.
type TurnManager struct {
	Deck       *entities.Deck
	Room       *Room
	WeaponSlot *entities.WeaponSlot

	HP              int
	AvoidedLastRoom bool
	LastCard        *entities.Card

	CarryOver     *entities.Card
	CarryOverSlot int // -1 if none

	GameOver bool
	Victory  bool
	Score    int

	Phase string // "ready", "dealing", "resolving", "room_complete"
}

// NewTurnManager creates a new game instance.
func NewTurnManager() *TurnManager {
	return &TurnManager{
		Deck:          entities.NewDeck(),
		Room:          NewRoom(),
		WeaponSlot:    entities.NewWeaponSlot(),
		HP:            core.StartHP,
		CarryOverSlot: -1,
		Phase:         "ready",
	}
}

// DealRoom deals cards into the room.
// Returns the slot cards array and the carry-over slot index (-1 if none).
func (tm *TurnManager) DealRoom() ([4]*entities.Card, int) {
	tm.Phase = "dealing"
	var slotCards [4]*entities.Card
	coSlot := -1

	// Carry-over stays in its original slot
	if tm.CarryOver != nil {
		coSlot = tm.CarryOverSlot
		slotCards[coSlot] = tm.CarryOver
		tm.CarryOver = nil
		tm.CarryOverSlot = -1
	}

	// Fill empty slots from the dungeon deck
	for i := 0; i < 4; i++ {
		if slotCards[i] == nil && !tm.Deck.IsEmpty() {
			slotCards[i] = tm.Deck.DrawCard()
		}
	}

	tm.Room.SetCards(slotCards)
	tm.Phase = "resolving"

	lib.EventEmit("room_dealt", map[string]any{
		"slot_cards":      slotCards,
		"carry_over_slot": coSlot,
	})
	return slotCards, coSlot
}

// CanAvoid returns true if the player can skip the current room.
func (tm *TurnManager) CanAvoid() bool {
	return CanAvoidRoom(tm.AvoidedLastRoom) &&
		tm.Room.CardCount() == 4 &&
		tm.Phase == "resolving" &&
		tm.Room.ResolvedCount == 0
}

// AvoidRoom sends all 4 room cards to the bottom of the deck.
func (tm *TurnManager) AvoidRoom() bool {
	if !tm.CanAvoid() {
		return false
	}

	cards := tm.Room.GetAllCards()
	tm.Room.Clear()
	tm.Deck.PlaceAtBottom(cards)
	tm.AvoidedLastRoom = true

	lib.EventEmit("room_avoided", map[string]any{
		"cards": cards,
	})
	return true
}

// ResolveCard resolves a card at the given slot index.
func (tm *TurnManager) ResolveCard(slotIndex int, useWeapon bool) *ResolveResult {
	if tm.Phase != "resolving" {
		return nil
	}
	if !tm.Room.HasCardAt(slotIndex) {
		return nil
	}

	card := tm.Room.Slots[slotIndex]
	var result *ResolveResult

	if card.IsPotion {
		result = tm.resolvePotion(slotIndex, card)
	} else if card.IsWeapon {
		result = tm.resolveWeapon(slotIndex, card)
	} else if card.IsMonster {
		result = tm.resolveMonster(slotIndex, card, useWeapon)
	}

	if tm.Room.IsComplete() {
		tm.finishRoom()
	}

	if tm.HP <= 0 {
		tm.endGame()
	}

	return result
}

func (tm *TurnManager) resolvePotion(slotIndex int, card *entities.Card) *ResolveResult {
	healed := 0
	wasted := false

	if CanUsePotion(tm.Room) {
		oldHP := tm.HP
		tm.HP = ApplyHealing(tm.HP, card.Value, core.MaxHP)
		healed = tm.HP - oldHP
		tm.Room.PotionUsed = true
	} else {
		wasted = true
	}

	tm.Room.ResolveCard(slotIndex)
	tm.LastCard = card

	result := &ResolveResult{
		Action: "potion",
		Card:   card,
		Slot:   slotIndex,
		Healed: healed,
		Wasted: wasted,
	}
	lib.EventEmit("potion_used", result)
	return result
}

func (tm *TurnManager) resolveWeapon(slotIndex int, card *entities.Card) *ResolveResult {
	oldWeapon, oldSlain := tm.WeaponSlot.Equip(card)
	tm.Room.ResolveCard(slotIndex)
	tm.LastCard = card

	result := &ResolveResult{
		Action:   "weapon_equipped",
		Card:     card,
		Slot:     slotIndex,
		OldWeapon: oldWeapon,
		OldSlain:  oldSlain,
	}
	lib.EventEmit("weapon_equipped", result)
	return result
}

func (tm *TurnManager) resolveMonster(slotIndex int, card *entities.Card, useWeapon bool) *ResolveResult {
	var damage int
	usedWeapon := false

	if useWeapon && CanUseWeapon(tm.WeaponSlot, card) {
		damage = WeaponDamage(card.Value, tm.WeaponSlot.WeaponValue())
		tm.WeaponSlot.RecordKill(card)
		usedWeapon = true
	} else {
		damage = BarehandedDamage(card.Value)
	}

	tm.HP = ApplyDamage(tm.HP, damage)
	tm.Room.ResolveCard(slotIndex)
	tm.LastCard = card

	result := &ResolveResult{
		Action:     "monster_fought",
		Card:       card,
		Slot:       slotIndex,
		Damage:     damage,
		UsedWeapon: usedWeapon,
	}
	lib.EventEmit("monster_fought", result)
	return result
}

func (tm *TurnManager) finishRoom() {
	carryOver, coSlot := tm.Room.GetRemainingCard()
	tm.CarryOver = carryOver
	tm.CarryOverSlot = coSlot
	tm.Room.Clear()
	tm.AvoidedLastRoom = false

	tm.Phase = "room_complete"
	lib.EventEmit("room_complete", map[string]any{
		"carry_over":      tm.CarryOver,
		"carry_over_slot": tm.CarryOverSlot,
	})

	if tm.Deck.IsEmpty() && tm.CarryOver == nil {
		tm.endGame()
	}
}

// EndGame ends the game and calculates the score.
func (tm *TurnManager) EndGame() {
	tm.endGame()
}

func (tm *TurnManager) endGame() {
	if tm.GameOver {
		return
	}
	tm.GameOver = true
	tm.Victory = tm.HP > 0
	tm.Score = ScoreCalculate(tm.HP, tm.Deck, tm.LastCard)

	lib.EventEmit("game_over", map[string]any{
		"victory": tm.Victory,
		"score":   tm.Score,
		"hp":      tm.HP,
	})
}

// GetActions returns available actions for a card at the given slot.
func (tm *TurnManager) GetActions(slotIndex int) []string {
	if tm.Phase != "resolving" {
		return nil
	}
	if !tm.Room.HasCardAt(slotIndex) {
		return nil
	}

	card := tm.Room.Slots[slotIndex]
	var actions []string

	if card.IsPotion {
		actions = append(actions, "use_potion")
	} else if card.IsWeapon {
		actions = append(actions, "equip_weapon")
	} else if card.IsMonster {
		actions = append(actions, "fight_barehanded")
		if CanUseWeapon(tm.WeaponSlot, card) {
			actions = append(actions, "fight_with_weapon")
		}
	}
	return actions
}
